src/client.toit

```diff
@@ -95,35 +98,41 @@ class Client:
       throw "connection refused: $ack.return_code"
 
   /**
-  Close the MQTT Client.
+  Close the MQTT client.
   */
   close:
-    // TODO(anders): This can block, fix me.
-    incoming_.send null
-    try:
-      send_ DisconnectPacket
-    finally:
+    if is_closed: return
+    // We need to be able to close even when canceled, so we run the
+    // close steps in a critical region.
+    critical_do:
+      // TODO(anders): The incoming buffer can be full in which case this will
+      // block. All we want to achieve is to unblock the corresponding call to
+      // receive, so the task stuck in $handle can continue.
+      incoming_.send null
+      catch --trace=(: should_trace_exception_ it):
+        send_ DisconnectPacket
       pending_.do --values: it.set null
-      if task_:
-        task_.cancel
-        task_ = null
+      if task_: task_.cancel
 
-  send_ packet/Packet:
-    transport_.send packet
-    last_sent_us_ = Time.monotonic_us
+  /**
+  Returns whether or not the client is closed.
+  */
+  is_closed -> bool:
+    return next_packet_id_ == null
 
   /**
-  Publish a MQTT message on $topic.
+  Publish an MQTT message on $topic.
```

Comment:
```suggestion
  Publishes an MQTT message on $topic.
  
  The $qos parameter must be either:
  - 0: at most once, aka "fire and forget". In this configuration the message is sent, but the delivery
        is not guaranteed.
  - 1: at least once. The MQTT client ensures that the message is received by the MQTT broker.
  
  QOS = 2 (exactly once) is not implemented by this client.
  
  The $retain parameter lets the MQTT broker know whether it should retain this message. A new (later)
    subscription to this $topic would receive the retained message, instead of needing to wait for
    a new message on that topic.
```